internal/pgproxy/pgproxy.go

```diff
@@ -0,0 +1,194 @@
+package pgproxy
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/jackc/pgx/v5/pgconn"
+	"github.com/jackc/pgx/v5/pgproto3"
+
+	"github.com/TBD54566975/ftl/internal/log"
+)
+
+type PgProxy struct {
+	// listenAddress is the address to listen on for incoming connections.
+	listenAddress string
+	// connectionStringFn is a function that constructs a new connection string from parameters of the incoming connection.
+	connectionStringFn func(ctx context.Context, params map[string]string) (string, error)
+}
+
+// NewPgProxy creates a new PgProxy.
+func NewPgProxy(address string, connectionFn func(ctx context.Context, params map[string]string) (string, error)) *PgProxy {
+	return &PgProxy{
+		listenAddress:      address,
+		connectionStringFn: connectionFn,
+	}
+}
+
+// Start starts the proxy.
+func (p *PgProxy) Start(ctx context.Context) error {
+	logger := log.FromContext(ctx)
+
+	if p.connectionStringFn == nil {
+		return fmt.Errorf("ConnectionStringFn is required")
```

Comment:
I'd probably just omit these checks, but if you do want to keep them just panic - this is a misuse of the API and programmer error

Reply:
Ah, this was leftover before I added `New` Will remove, as it is not needed anymore